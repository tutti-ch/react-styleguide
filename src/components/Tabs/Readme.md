### Tabs component

You can create tabbed views using this this component.

```jsx
<Tabs>
  <Tabs.Tab title="This is a tab">
    <h1>An amazing tab</h1>
    <div>lots of stuff</div>
  </Tabs.Tab>

  <Tabs.Tab title="This is another tab">
    <h1>Another amazing tab</h1>
    <img src="http://www.cryptolulz.com/content/uploads/images/December2017/me-meme.jpg" />
  </Tabs.Tab>

  <Tabs.Tab title="And they keep coming">
    <h1>Can do magic here</h1>
    <div>lots of stuff</div>
    <img src="https://i.redd.it/c47izks1x8gz.jpg" />
  </Tabs.Tab>

  <Tabs.Tab title="Secret ICO">
    <h1>DOGE</h1>
    <div>lots of stuff</div>
    <div>much WOW</div>
  </Tabs.Tab>
</Tabs>;
```
