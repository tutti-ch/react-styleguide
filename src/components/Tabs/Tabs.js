import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import classes from "./Tabs.scss";
import Tab from "./_Tab";

export default class Tabs extends Component {
  static defaultProps = { defaultActiveTab: 0 };
  static Tab = Tab;

  constructor(props, context) {
    super(props, context);
    this.state = { activeTab: this.props.defaultActiveTab };

    this.onTabClick = this.onTabClick.bind(this);
  }

  /**
   * Handle the tab being clicked.
   *
   * @param {*} tab
   */
  onTabClick(tab) {
    this.setState({ activeTab: tab });
  }

  /**
   * Takes the children passed to Tabs and enhance them with some extra props
   * to keep track of their state
   */
  renderTabs() {
    return React.Children.map(this.props.children, (child, index) => {
      if (!child) return null;

      return React.cloneElement(child, {
        index,
        onClick: this.onTabClick,
        isActive: index === this.state.activeTab
      });
    });
  }

  /**
   * If defined it renders the content of the current active children
   */
  renderActiveTabContent() {
    const { children } = this.props;
    const { activeTab } = this.state;
    if (children[activeTab]) {
      return children[activeTab].props.children;
    }
  }

  render() {
    const { className, children } = this.props;
    const tabClasses = classNames(className, [classes.tabs]);

    return (
      <div className={tabClasses}>
        {this.renderTabs()}
        <div className={classes.content}>{this.renderActiveTabContent()}</div>
      </div>
    );
  }
}
