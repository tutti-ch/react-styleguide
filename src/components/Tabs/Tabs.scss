@import "../../styles/main";

$border-color: $greyscale-85;

.tabs {
  width: 100%;
}

.tab {
  user-select: none;
  display: inline-block;
  margin: 0 -1px -1px -1px;
  padding: $size-m;
  text-align: center;
  background: $greyscale-95;
  border: 1px solid $border-color;

  transition: all $transition-m;

  &.active {
    background: $greyscale-100;
    border: 1px solid $border-color;
    border-bottom: 1px solid $greyscale-100;
  }

  &:hover {
    cursor: pointer;
  }
}

.content {
  width: 100%;
  display: block;
  border-top: 1px solid $border-color;
}
