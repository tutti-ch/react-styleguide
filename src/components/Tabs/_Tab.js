import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import classes from "./Tabs.scss";

export default class Tab extends Component {
  constructor(props) {
    super(props);

    this.handleOnClick = this.handleOnClick.bind(this);
  }

  handleOnClick(e) {
    const { onClick, index } = this.props;
    onClick(index);
  }

  render() {
    const { isActive, title } = this.props;
    const tabClasses = classNames(classes.tab, { [classes.active]: isActive });

    return (
      <div className={tabClasses} onClick={this.handleOnClick}>
        {title}
      </div>
    );
  }
}
