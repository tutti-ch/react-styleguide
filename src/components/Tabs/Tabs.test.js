/* global describe, expect, test, jest */
import React from "react";
import { mount } from "enzyme";
import Tabs from "./index";

describe("(Component) Tabs", () => {
  const tree = mount(
    <Tabs>
      <Tabs.Tab title="This is a tab">
        <h1>An amazing tab</h1>
        <div>lots of stuff</div>
      </Tabs.Tab>

      <Tabs.Tab title="This is another tab">
        <h1>Another amazing tab</h1>
        <img src="http://www.cryptolulz.com/content/uploads/images/December2017/me-meme.jpg" />
      </Tabs.Tab>

      <Tabs.Tab title="And they keep coming">
        <h1>Can do magic here</h1>
        <div>lots of stuff</div>
        <img src="https://i.redd.it/c47izks1x8gz.jpg" />
      </Tabs.Tab>

      <Tabs.Tab title="Secret ICO">
        <h1>DOGE</h1>
        <div>lots of stuff</div>
        <div>much WOW</div>
      </Tabs.Tab>
    </Tabs>
  );

  test("should render correctly", () => {
    expect(tree).toMatchSnapshot();
  });

  test("should change tab when clicked", () => {
    expect(tree.state("activeTab")).toBe(0);
    tree
      .find("Tab")
      .at(2)
      .simulate("click");
    expect(tree.state("activeTab")).toBe(2);
  });
});
