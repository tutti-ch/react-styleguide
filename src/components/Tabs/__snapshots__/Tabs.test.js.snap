// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Tabs should render correctly 1`] = `
<Tabs
  defaultActiveTab={0}
>
  <div
    className=""
  >
    <Tab
      index={0}
      isActive={true}
      key=".0"
      onClick={[Function]}
      title="This is a tab"
    >
      <div
        className="undefined"
        onClick={[Function]}
      >
        This is a tab
      </div>
    </Tab>
    <Tab
      index={1}
      isActive={false}
      key=".1"
      onClick={[Function]}
      title="This is another tab"
    >
      <div
        className=""
        onClick={[Function]}
      >
        This is another tab
      </div>
    </Tab>
    <Tab
      index={2}
      isActive={false}
      key=".2"
      onClick={[Function]}
      title="And they keep coming"
    >
      <div
        className=""
        onClick={[Function]}
      >
        And they keep coming
      </div>
    </Tab>
    <Tab
      index={3}
      isActive={false}
      key=".3"
      onClick={[Function]}
      title="Secret ICO"
    >
      <div
        className=""
        onClick={[Function]}
      >
        Secret ICO
      </div>
    </Tab>
    <div>
      <h1>
        An amazing tab
      </h1>
      <div>
        lots of stuff
      </div>
    </div>
  </div>
</Tabs>
`;
